src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/AbstractDocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeFixes
+{
+    /// <summary>
+    /// Provides a base class to write a <see cref="FixAllProvider"/> that fixes documents independently.
+    /// </summary>
+    internal abstract class AbstractDocumentBasedFixAllProvider : FixAllProvider
+    {
+        protected abstract string GetFixAllTitle(FixAllContext fixAllContext);
+
+        protected abstract Task<Document?> FixAllAsync(FixAllContext context, Document document, ImmutableArray<Diagnostic> diagnostics);
+
+#pragma warning disable RS0005 // Do not use generic 'CodeAction.Create' to create 'CodeAction'
+
+        public sealed override Task<CodeAction?> GetFixAsync(FixAllContext fixAllContext)
+        {
+            var title = GetFixAllTitle(fixAllContext);
+
+            switch (fixAllContext.Scope)
+            {
+                case FixAllScope.Document:
+                    return Task.FromResult<CodeAction?>(CodeAction.Create(
+                        title,
+                        c => GetDocumentFixesAsync(fixAllContext.WithCancellationToken(c)),
+                        nameof(DocumentBasedFixAllProvider)));
+
+                case FixAllScope.Project:
+                    return Task.FromResult<CodeAction?>(CodeAction.Create(
+                        title,
+                        c => GetProjectFixesAsync(fixAllContext.WithCancellationToken(c), fixAllContext.Project),
+                        nameof(DocumentBasedFixAllProvider)));
+
+                case FixAllScope.Solution:
+                    return Task.FromResult<CodeAction?>(CodeAction.Create(
+                        title,
+                        c => GetSolutionFixesAsync(fixAllContext.WithCancellationToken(c)),
+                        nameof(DocumentBasedFixAllProvider)));
+
+                case FixAllScope.Custom:
+                default:
+                    return Task.FromResult<CodeAction?>(null);
```

Comment:
this code (and the majority of this file) was moved up from DoucmentBasedFixAllProvider).